File: services/buffmarket.go
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/cyberbebebe/cs2-profit-checker/types"
	"github.com/cyberbebebe/cs2-profit-checker/utils"
)

type BuffMarketService struct {
	Headers map[string]string
	Cookies map[string]string
	Client *http.Client
}

func NewBuffMarketService(headers, cookies map[string]string) *BuffMarketService {
    return &BuffMarketService{
		Headers: headers,
		Cookies: cookies,
        Client: &http.Client{
            Timeout: 10 * time.Second,
        },
    }
}

func (s *BuffMarketService) Name() string {
    return "BuffMarket"
}

func (s *BuffMarketService) GetSales(startTime, endTime time.Time)([]types.Transaction, error){
	return s.fetchHistory("sell", startTime, endTime)
}

func (s *BuffMarketService) GetBuys(startTime, endTime time.Time)([]types.Transaction, error){
	return s.fetchHistory("buy", startTime, endTime)
}


func (s *BuffMarketService) fetchHistory(action string, startTime, endTime time.Time) ([]types.Transaction, error){
	var allTxs []types.Transaction
	page := 1

	for{
		url := fmt.Sprintf("https://api.buff.market/api/market/%s_order/history?game=csgo&page_num=%d&page_size=200&state=success", action, page)

		req, _ := http.NewRequest("GET", url, nil)

		for k, v := range s.Headers {
			req.Header.Set(k, v)
		}
		for k, v := range s.Cookies {
			req.AddCookie(&http.Cookie{Name: k, Value: v})
		}

		resp, err := s.Client.Do(req)
		if err != nil{
			return nil, fmt.Errorf("request failed: %v", err)
		}
		defer resp.Body.Close()

		if resp.StatusCode != 200{
			return nil, fmt.Errorf("BuffMarket response error: %v", err)
		}

		var response buffMarketResponse
		if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
			return nil, fmt.Errorf("json decode error: %v", err)
		}

		if response.Code != "OK" {
			return nil, fmt.Errorf("buff API error: %s - %s", response.Code, response.Msg)
		}

		for _, raw := range response.Data.Items{

			if raw.State != "SUCCESS"{
				continue
			}

			goodsIDStr := strconv.Itoa(raw.GoodsID)
			info, exists := response.Data.GoodsInfos[goodsIDStr]

			itemName := "Unknown Item"
			if exists {
				itemName = info.ItemName
			}

			cleanTx := convertBuffMarketTx(raw, itemName, action)
			
			if cleanTx.Date.After(endTime) {
				continue
			}
			
			if cleanTx.Date.Before(startTime) {
				return allTxs, nil
			}

			allTxs = append(allTxs, cleanTx)

		}

		if page >= response.Data.TotalPages{
			return allTxs, nil
		}

		page++

		time.Sleep(2 * time.Second)
	}

}

func convertBuffMarketTx(raw buffItemObject, itemName string, action string) types.Transaction{

	txType := types.TxSell
	if action == "buy" {
		txType = types.TxBuy
	}

	priceVal, _ := strconv.ParseFloat(raw.Price, 64)
	feeVal, _ := strconv.ParseFloat(raw.Fee, 64)

	finalPrice := priceVal - feeVal

	floatVal, _ := strconv.ParseFloat(raw.AssetInfo.Paintwear, 64)
	date := time.Unix(raw.UpdatedAt, 0)

	finalPattern := -1

	if raw.AssetInfo.Info.Paintseed != nil{
		finalPattern = *raw.AssetInfo.Info.Paintseed
	}
	
	// Charm check
	if len(raw.AssetInfo.Info.Keychains) > 0 {
		charmObject := raw.AssetInfo.Info.Keychains[0]
		
		// Contains for situations where "Charm" part is not in CharmName
		if strings.Contains(itemName, charmObject.CharmName) {
			finalPattern = *charmObject.CharmPattern
    	}
	}

	tx := types.Transaction{
		Source:    "BuffMarket",
		Type:      txType,
		TxID:      raw.TxID,
		AssetID:   raw.AssetInfo.AssetID,
		ItemName:  itemName,
		Price:     finalPrice,
		Currency:  "USD",
		Date:      date,
		FloatVal:  floatVal,
		Pattern:   finalPattern,
		Phase:     raw.AssetInfo.Info.Metaphysic.Data.Phase,
	}

	tx.Signature = utils.GenerateSignature(tx.ItemName, tx.FloatVal, finalPattern)
	
	return tx
}

File: js/fetchers/buffmarket.js
import { BaseFetcher } from "./base.js";
import { Transaction } from "./base.js";

export class BuffMarketFetcher extends BaseFetcher {
  constructor() {
    super("BuffMarket");
  }

  async checkSession() {
    try {
      const data = await this.fetchWithAuth(
        "https://api.buff.market/account/api/login/status",
      );
      if (data && data.code === "OK" && data.data && data.data.state === 2) {
        return true;
      }
      return false;
    } catch (e) {
      return false;
    }
  }
  async getBalance() {
    try {
      const url = "https://api.buff.market/api/asset/get_brief_asset/";
      const data = await this.fetchWithAuth(url);

      if (!data) return { amount: 0, currency: "USD" };

      const usd = parseFloat(data.data.total_amount || 0);
      const frozen = parseFloat(data.data.frozen_amount || 0);
      const pending = parseFloat(data.data.pending_divide_amount || 0);

      return { amount: usd + frozen + pending, currency: "USD" };
    } catch (e) {
      console.error("[BuffMarket] Balance error:", e);
      return { amount: 0, currency: "USD" };
    }
  }

  async getSales() {
    return this.getHistory("sell");
  }

  async getBuys() {
    return this.getHistory("buy");
  }

  async getHistory(type) {
    let allTxs = [];
    let page = 1;
    const limit = 200;
    const orderType = type === "sell" ? "sell_order" : "buy_order";

    while (true) {
      const url = `https://api.buff.market/api/market/${orderType}/history?game=csgo&page_num=${page}&page_size=${limit}&state=success`;

      let data;
      try {
        data = await this.fetchWithAuth(url);
      } catch (e) {
        console.error(`[BuffMarket] Error fetching page ${page}:`, e);
        break;
      }

      if (data.code !== "OK") {
        console.error(`[BuffMarket] API Error: ${data.msg}`);
        break;
      }

      const items = data.data.items || [];
      const goodsInfos = data.data.goods_infos || {};

      if (items.length === 0) break;

      for (const raw of items) {
        if (raw.state !== "SUCCESS") continue;

        const goodsId = raw.goods_id;
        const info = goodsInfos[goodsId];
        const itemName = info ? info.market_hash_name : "Unknown Item";

        // Conversion logic
        const priceVal = parseFloat(raw.price || 0);
        const feeVal = parseFloat(raw.fee || 0);

        // finalPrice := priceVal - feeVal
        const finalPrice = priceVal - feeVal;

        // Date: updated_at is Unix timestamp (seconds)
        const date = new Date(raw.updated_at * 1000);

        // Metadata
        const assetInfo = raw.asset_info || {};
        const paintwear = parseFloat(assetInfo.paintwear || 0);

        const infoObj = assetInfo.info || {};
        let finalPattern =
          infoObj.paintseed !== undefined ? infoObj.paintseed : -1;

        // Phase / Doppler
        const phase = infoObj.metaphysic?.data?.name || "";

        if (infoObj.keychains && infoObj.keychains.length > 0) {
          const charm = infoObj.keychains[0];
          if (charm && itemName.includes(charm.name)) {
            finalPattern = charm.pattern;
          }
        }
        const createdDate = raw.created_at
          ? new Date(raw.created_at * 1000)
          : new Date();
        const verifiedDate = raw.updated_at
          ? new Date(raw.updated_at * 1000)
          : null;
        allTxs.push(
          new Transaction({
            source: "BuffMarket",
            type: type === "sell" ? "SELL" : "BUY",
            tx_id: raw.id,
            asset_id: assetInfo.assetid,
            item_name: itemName,
            price: parseFloat(finalPrice.toFixed(2)),
            currency: "USD", // Buff Market uses USD
            created_at: createdDate,
            verified_at: verifiedDate,
            float_val: paintwear,
            pattern: finalPattern,
            phase: phase,
          }),
        );
      }

      // Pagination Check
      // Go: if page >= response.Data.TotalPages
      if (page >= data.data.total_page) {
        break;
      }

      page++;
      await this.sleep(500);
    }

    return allTxs;
  }
}
